contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+
+The schema types follow those of the OpenAPI library, but we decided to define them independently for the API to have full control over the serialization and versioning. Hence, it is easy to convert our types into those used for validation or to integrate them into an OpenAPI spec later.
+
+Reference http://json-schema.org is also used by OpenAPI. We propose this as there are implementations available in Go and with OpenAPI, we will also be able to serve OpenAPI specs for CustomResourceDefintions.
+
+```go
+// CustomResourceSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+    Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
+    Version string `json:"version" protobuf:"bytes,2,opt,name=version"`
+    Names CustomResourceDefinitionNames `json:"names" protobuf:"bytes,3,opt,name=names"`
+    Scope ResourceScope `json:"scope" protobuf:"bytes,8,opt,name=scope,casttype=ResourceScope"`
+    // Validation describes the validation methods for CustomResources
+    Validation CustomResourceValidation `json:"validation,omitempty"`
+}
+
+// CustomResourceValidation is a list of validation methods for CustomResources
+type CustomResourceValidation struct {
+    // JSONSchema is the JSON Schema to be validated against.
+    // Can add other validation methods later if needed.
+    JSONSchema *JSONSchemaProps `json:"jsonSchema,omitempty"`
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
+	ID                   string                     `json:"id,omitempty"`
+	Schema               JSONSchemaURL              `json:"-,omitempty"`
+	Ref                  JSONSchemaRef              `json:"-,omitempty"`
+	Description          string                     `json:"description,omitempty"`
+	Type                 StringOrArray              `json:"type,omitempty"`
+	Format               string                     `json:"format,omitempty"`
+	Title                string                     `json:"title,omitempty"`
+	Default              interface{}                `json:"default,omitempty"`
+	Maximum              *float64                   `json:"maximum,omitempty"`
+	ExclusiveMaximum     bool                       `json:"exclusiveMaximum,omitempty"`
+	Minimum              *float64                   `json:"minimum,omitempty"`
+	ExclusiveMinimum     bool                       `json:"exclusiveMinimum,omitempty"`
+	MaxLength            *int64                     `json:"maxLength,omitempty"`
+	MinLength            *int64                     `json:"minLength,omitempty"`
+	Pattern              string                     `json:"pattern,omitempty"`
+	MaxItems             *int64                     `json:"maxItems,omitempty"`
+	MinItems             *int64                     `json:"minItems,omitempty"`
+	// disable uniqueItems for now because it can cause the validation runtime
+	// complexity to become quadratic.
```

Comment:
CRDs are inherently a cluster-admin resource since they affect the resources for the entire cluster.  I don't feel bad about allowing a cluster-admin to make an expensive resource.